Sources/Controllers/Map/Toolbars/OAWeatherToolbar.mm

```diff
@@ -0,0 +1,465 @@
+//
+//  OAWeatherToolbar.mm
+//  OsmAnd
+//
+//  Created by Skalii on 03.06.2022.
+//  Copyright (c) 2022 OsmAnd. All rights reserved.
+//
+
+#import "OAWeatherToolbar.h"
+#import "OARootViewController.h"
+#import "OAMapHudViewController.h"
+#import "OAWeatherLayerSettingsViewController.h"
+#import "OASegmentedSlider.h"
+#import "OAMapStyleSettings.h"
+#import "OAMapLayers.h"
+#import "Localization.h"
+
+#define kTempIndex 0
+#define kPressureIndex 1
+#define kWindIndex 2
+#define kCloudIndex 3
+#define kPrecipitationIndex 4
+#define kContoursIndex 5
+
+@implementation OAWeatherToolbarLayersDelegate
+{
+    OAMapStyleSettings *_styleSettings;
+    OsmAndAppInstance _app;
+
+    NSMutableArray<NSMutableDictionary *> *_data;
+}
+
+- (instancetype)init
+{
+    self = [super init];
+    if (self)
+    {
+        [self commonInit];
+    }
+    return self;
+}
+
+- (void)commonInit
+{
+    _styleSettings = [OAMapStyleSettings sharedInstance];
+    _app = [OsmAndApp instance];
+
+    [self updateData];
+}
+
+- (void)updateData
+{
+    NSMutableArray<NSMutableDictionary *> *layersData = [NSMutableArray array];
+    [layersData addObject:[NSMutableDictionary dictionaryWithDictionary:@{
```

Comment:
Create a simple private class to hold the state rather than copying numerous dictionaries in memory